src/FreetypeRenderer.cc

```diff
@@ -256,49 +319,186 @@ std::vector<const Geometry *> FreetypeRenderer::render(const FreetypeRenderer::P
 			continue;
 		}
 		const GlyphData *glyph_data = new GlyphData(glyph, idx, &glyph_pos[idx]);
 		glyph_array.push_back(glyph_data);
 	}
 
-	double width = 0, ascend = 0, descend = 0;
-	for (const auto glyph : glyph_array) {
-		FT_BBox bbox;
-		FT_Glyph_Get_CBox(glyph->get_glyph(), FT_GLYPH_BBOX_GRIDFIT, &bbox);
-		
-		if (HB_DIRECTION_IS_HORIZONTAL(hb_buffer_get_direction(hb_buf))) {
-			double asc = std::max(0.0, bbox.yMax * unscale);
-			double desc = std::max(0.0, -bbox.yMin * unscale);
-			width += glyph->get_x_advance() * params.spacing;
-			ascend = std::max(ascend, asc);
-			descend = std::max(descend, desc);
-		} else {
-			double w_bbox = (bbox.xMax - bbox.xMin) * unscale;
-			width = std::max(width, w_bbox);
-			ascend += glyph->get_y_advance() * params.spacing;
-		}
-	}
-	
-	double x_offset = calc_x_offset(params.halign, width);
-	double y_offset = calc_y_offset(params.valign, ascend, descend);
+    ascent = std::numeric_limits<double>::lowest();
+    descent = std::numeric_limits<double>::max();
+    advance_x = 0;
+    advance_y = 0;
+    left = std::numeric_limits<double>::max();
+    right = std::numeric_limits<double>::lowest();
+    bottom = std::numeric_limits<double>::max();
+    top = std::numeric_limits<double>::lowest();
+
+    bool has_ink = false;
+    for (const auto glyph : glyph_array) {
+        FT_BBox bbox;
+        FT_Glyph_Get_CBox(glyph->get_glyph(), FT_GLYPH_BBOX_GRIDFIT, &bbox);
+
+        // Note that glyphs can extend left of their origin
+        // and right of their advance-width, into the next
+        // glyph's box.  In theory they could extend
+        // arbitrarily through other glyphs' boxes.  Hence the
+        // need for the min/max here.
+
+        // Glyphs with null bounding boxes do not contribute
+        // ink and so do not contribute to the bounding box or
+        // ascent and descent.
+        if (bbox.xMax > bbox.xMin && bbox.yMax > bbox.yMin) {
+            ascent = std::max(ascent, bbox.yMax / scale);
+            descent = std::min(descent, bbox.yMin / scale);
+
+            double gxoff = glyph->get_x_offset();
+            double gyoff = glyph->get_y_offset();
+
+            left = std::min(left,
+                advance_x + gxoff + bbox.xMin/scale);
+            right = std::max(right,
+                advance_x + gxoff + bbox.xMax/scale);
+
+            top = std::max(top,
+                advance_y + gyoff + bbox.yMax/scale);
+            bottom = std::min(bottom,
+                advance_y + gyoff + bbox.yMin/scale);
+
+            has_ink = true;
+        }
+
+        advance_x += glyph->get_x_advance() * params.spacing;
+        advance_y += glyph->get_y_advance() * params.spacing;
+    }
+
+    if (has_ink) {
+        if (HB_DIRECTION_IS_HORIZONTAL(
+            hb_buffer_get_direction(hb_buf))) {
+            calc_offsets_horiz(params);
```

Comment:
Maybe we could use one of the Eigen library types to return x/y instead of having `calc_offsets_*` set the values directly? That way those could stay `const` making it slightly easier to reason about the data flow.

Reply:
I don't think so, or at least not without making it so some of the shaping results are returned as the return value from `shape()`, while others are returned as members of the `ShapeResults` object.  (Or maybe I'm just not understanding.)

Note that `calc_offsets_*()` are private functions called only once.  Their reason for existence is purely to factor out the sequences that calculate those offsets as readily identifiable components.